pydantic/_internal/_generics.py

```diff
@@ -0,0 +1,327 @@
+
+    # Handle special case for typehints that can have lists as arguments.
+    # `typing.Callable[[int, str], int]` is an example for this.
+    if isinstance(type_, (List, list)):
+        resolved_list = list(replace_types(element, type_map) for element in type_)
+        if all_identical(type_, resolved_list):
+            return type_
+        return resolved_list
+
+    if isinstance(type_, PydanticForwardRef):
+        # queue the replacement as a deferred action
+        return type_.replace_types(type_map)
+
+    # If all else fails, we try to resolve the type directly and otherwise just
+    # return the input with no modifications.
+    return type_map.get(type_, type_)
+
+
+def check_parameters_count(cls: type[BaseModel], parameters: tuple[Any, ...]) -> None:
+    actual = len(parameters)
+    expected = len(cls.__pydantic_generic_parameters__ or ())
+    if actual != expected:
+        description = 'many' if actual > expected else 'few'
+        raise TypeError(f'Too {description} parameters for {cls}; actual {actual}, expected {expected}')
+
+
+_visit_counts_context: ContextVar[dict[str, int] | None] = ContextVar('_visit_counts_context', default=None)
+
+
+@contextmanager
+def generic_recursion_self_type(origin: type[BaseModel], args: tuple[Any, ...]) -> Iterator[PydanticForwardRef | None]:
+    """
+    This contextmanager should be placed around recursive calls used to build a generic type,
+    and accept as arguments the generic origin type and the type arguments being passed to it.
+
+    If the same origin and arguments are observed twice, it implies that a self-reference placeholder
+    can be used while building the core schema, and will produce a schema_ref that will be valid in the
+    final parent schema.
+
+    I believe the main reason that the same origin/args must be observed twice is that a BaseModel's
+    inner_schema will be a TypedDictSchema that doesn't include the first occurrence of the PydanticForwardRef
+    reference, so the referenced schema may not end up in the final core_schema unless you expand two
+    layers deep.
+    """
+    visit_counts_by_ref = _visit_counts_context.get()
+    if visit_counts_by_ref is None:
+        visit_counts_by_ref = defaultdict(int)
+        token = _visit_counts_context.set(visit_counts_by_ref)
+    else:
+        token = None
+
+    type_ref = get_type_ref(origin, args_override=args)
+    if visit_counts_by_ref[type_ref] >= 2:
+        self_type = PydanticForwardRef(
+            core_schema.definition_reference_schema(type_ref), origin, ({'kind': 'class_getitem', 'item': args},)
+        )
+        yield self_type
+    else:
+        visit_counts_by_ref[type_ref] += 1
+        yield None
+    if token:
+        _visit_counts_context.reset(token)
+
+
+def recursively_defined_type_refs() -> set[str]:
+    return set((_visit_counts_context.get() or {}).keys())
+
+
+_GENERIC_TYPES_CACHE = GenericTypesCache()
+
+
+def _union_orderings_key(typevar_values: Any) -> Any:
```

Comment:
I did some digging into what was going on with the calls to get_args in the v1 cache key logic, and as far as I could tell the only thing it was really doing for us was helping us differentiate between the ordering of items in Unions (e.g., `Union[int, float]` vs. `Union[float, int]`). (It's definitely possible I'm missing other cases where it's relevant.)

I have pulled out this function to do exactly this check, and more "correctly" by applying to each item in a possible tuple (which is not currently done in v1), and ignoring the args for other types (besides Union) where order is not ignored and therefore not important (and differences in the result of get_args for other origin types can cause spurious cache misses).

Let me know if you know of any cases that also need handling besides when the origin is Union.

Reply:
humm, but union or der can be important in pydantic, are we sure it's correct to ignore order?

Reply:
I don't think this should block merging - we can reconsider this later.

Reply:
This is handling things the same as v1 currently, as far as I can tell. If something is now mishandled, I think it will be related to arguments in _non_-union types. There are no unit tests indicating this can cause issues so I'm inclined to leave like this for now but we can obviously fix it if we discover a use case that isn't addressed (and it would be good to add a unit test for it anyway).